### PR TITLE
PLAT-53627: Consolidate ProgressBarTooltip positioning props

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

We have two props to control positioning a `ProgressBarTooltip` which could be consolidated into one.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

* Consolidated into `side` which supports `before`, `after`, `left`, and `right`
* Converted `tooltip` to be either `bool`/`element`/`func` for consistency with the new Slider rewrite

Enact-DCO-1.0-Signed-off-by: Ryan Duffy (ryan.duffy@lge.com)

### CODE REVIEW

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
 import Tooltip from '../TooltipDecorator/Tooltip';
 
 import css from './ProgressBarTooltip.less';
 
+const memoizedPercentFormatter = memoize((/* locale */) => new NumFmt({
+	type: 'percentage',
+	useNative: false
+}));
+
+const getSide = (orientation, side) => {
```
Comment (teckliew) [inline packages/moonstone/ProgressBar/ProgressBarTooltip.js]: Should we also have a guard like `getOrientation` for `orientation` as well? When `orientation` gets  a non recognizable value, the tooltip doesn't position correctly.

Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: `PropTypes` should be a sufficient guard for that. In that light, we could (should?) move this logic into a custom PropType function instead which would make them consistent.

File: packages/moonstone/ProgressBar/ProgressBar.js [hunk 2]
```diff
@@ -69,45 +71,37 @@ const ProgressBarBase = kind({
...
 		 */
-		tooltipSide: PropTypes.string
+		tooltip: PropTypes.oneOfType([PropTypes.bool, PropTypes.object, PropTypes.func])
 	},
 
 	defaultProps: {
-		tooltip: false,
-		tooltipSide: 'before',
-		tooltipForceSide: false
+		tooltip: false
```
Comment (teckliew) [inline packages/moonstone/ProgressBar/ProgressBar.js]: Should we have `progress` default to 0?

**Comment by teckliew** [comment]
Looking awesome but a few loose ends:
Currently tooltip positions outside of the `ProgressBar` when `progress` is out of range; should we handle the tooltip position when `progress` is less than min and larger than max?
Not sure if the off by one error is in this PR or another one but I noticed that `IncrementSlider` skips `29%` and `58%`; and repeats `56%` twice

**Comment by teckliew** [comment]
About the `IncrementSlider`, it only reproduces in sampler for me 🤔

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 3]
```diff
@@ -94,30 +117,48 @@ const ProgressBarTooltipBase = kind({
 	contextTypes,
 
 	computed: {
-		className: ({forceSide, knobAfterMidpoint, orientation, side, styler}) => styler.append(orientation, {ignoreLocale: forceSide, knobAfterMidpoint}, side),
-		arrowAnchor: ({knobAfterMidpoint, orientation}) => {
+		children: ({children, proportion, percent}) => {
+			if (percent) {
+				const formatter = memoizedPercentFormatter(ilib.getLocale());
+
+				return formatter.format(Math.floor(proportion * 100));
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: Good catch. Running into some JS precision problems. Switching to `Math.round()` is a safer alternative.

**Comment by ryanjduffy** [comment]
Not too worried about `progress` being out of bounds. We already have dev mode warnings for the range in `ui/ProgressBar` and that's probably sufficient for now

**Comment by ryanjduffy** [comment]
Rebased this on `develop` since it previously included all the slider refactor commits.

**Pull request merged**